content/browser/attribution_reporting/aggregatable_attribution_utils.cc

```diff
@@ -121,6 +129,114 @@ std::vector<AggregatableHistogramContribution> CreateAggregatableHistogram(
   return contributions;
 }
 
+
+// TODO(kelly): Organize in classes
+void AttributionLogicLastTouch(Partition& partition, 
+        base::flat_map<std::string, std::vector<absl::uint128>>& trigger_keypieces_per_source) {
+
+  // for (uint64_t i=attribution_window.epoch_start(); i<=attribution_window.epoch_end(); i++) {
+  //   for (auto& pair : epoch_sources_count[i]) {
+  //     auto& source_key_piece = pair.first;
+  //     auto count = pair.second;
+
+  //     auto it = source_counts.find(source_key_piece);
+  //     if (it == source_counts.end()) {
+  //       source_counts[source_key_piece] = count;
+  //     }
+  //     source_counts[source_key_piece] += count;
+  //   }
+  // }
+}
+
+void AttributionLogicUniform(Partition& partition,
+        base::flat_map<std::string, std::vector<absl::uint128>>& trigger_keypieces_per_source) {
+
+  auto& attribution_window = partition.attribution_window;
+  auto& sources_per_epoch = partition.sources_per_epoch;
+  
+  double total_sources_count = 0;
+
+  base::flat_map<std::string, base::flat_map<absl::uint128, double>>
+          source_counts_per_sourcekey;
+
+
+  for (uint64_t i=attribution_window.epoch_start(); 
+          i <= attribution_window.epoch_end(); i++) {
+    
+    // Ignore empty epochs
+    auto it = sources_per_epoch.find(i);
+    if (it == sources_per_epoch.end()) {
+      continue;
+    }
+
+    // Count occurrences per source keypiece across all epochs
+    for (StoredSource* source : sources_per_epoch[i]) {
+      auto aggregation_keys = source->aggregation_keys().keys();
+      for (auto& pair : aggregation_keys) {
+        auto& source_key = pair.first;
+        auto& key_piece = pair.second;
+
+        auto it1 = source_counts_per_sourcekey.find(source_key);
+        if (it1 == source_counts_per_sourcekey.end()) {
+          source_counts_per_sourcekey[source_key] = {};
+        }
+
+        auto& source_counts = source_counts_per_sourcekey[source_key];
+        auto it2 = source_counts.find(key_piece);
+        if (it2 == source_counts.end()) {
+          source_counts[key_piece] = 0;
+        }
+        source_counts[key_piece] += 1;
+      }
+      total_sources_count++;
+    }
+  }
+
+  // Populate partition.report_value_pairs[*].report for all source_keys
+  for (auto& outer : source_counts_per_sourcekey) {
+    auto source_key = outer.first;
+    auto& report_value_pair = partition.report_value_pairs[source_key];
+    auto& trigger_keypieces = trigger_keypieces_per_source[source_key];
+
+    for (auto& inner : outer.second) {
+      auto source_keypiece = inner.first;
+      auto source_count = inner.second;
+
+      // Extend the source key_pieces for source_key
+      for (auto& trigger_keypiece : trigger_keypieces) {
+        source_keypiece |= trigger_keypiece;
+      }
+
+      double contribution_value = (source_count / total_sources_count) * report_value_pair.value;
+      report_value_pair.report.emplace_back(source_keypiece, contribution_value);      
```

Comment:
Ok so first we build `source_counts_per_sourcekey`, that is a list (actually dict) of sources, for each source we have a histogram that maps each keypiece to the number of times it appears in the whole attribution window. 

And next we browse all the sources. If there is a trigger associated to that source we concatenate the trigger_keypiece and map it to the right contribution value. But what if there is no trigger associated to that source? E.g. `source_key` does not appear in `trigger_keypieces_per_source`. Should we still give some value to the untouched `source_keypiece`? Maybe yes, it's a sort of unattributed impression?